Add `intersects?` to `Relation`

john-h-k opened the pull request:
Ruby 3.1 added `intersects?` which is equivalent to `(a & b).any?`. Rubocop added a corresponding cop, `Style/ArrayIntersect`, which transforms the old style to use `intersects?`. Unfortunately as `intersects?` is not delegated on `CollectionProxy`, this leads to false positives that need to be disabled for no good reason other than the fact the method isn't delegated.

This PR add delegation of `intersects?` to `Relation` which fixes this.

`intersection` and `union` are identical to `&` and `|` respectively (which are already delegated), so I thought for symmetry it was worth adding these, although I am not against removing them from this PR if it makes it more likely to be accepted.

<!--
Thanks for contributing to Rails!

Please do not make *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.

Create a pull request when it is ready for review and feedback
from the Rails team :).

If your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

About this template

The following template aims to help contributors write a good description for their pull requests.
We'd like you to provide a description of the changes in your pull request (i.e. bugs fixed or features added), motivation behind the changes, and complete the checklist below before opening a pull request.

Feel free to discard it if you need to (e.g. when you just fix a typo). -->

### Motivation / Background

<!--
Describe why this Pull Request needs to be merged. What bug have you fixed? What feature have you added? Why is it important?
If you are fixing a specific issue, include "Fixes #ISSUE" (replace with the issue number, remove the quotes) and the issue will be linked to this PR.
-->

This Pull Request has been created because [REPLACE ME]

### Detail

This Pull Request changes [REPLACE ME]

### Additional information

<!-- Provide additional information such as benchmarks, reference to other repositories or alternative solutions. -->

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

matthewd commented:
> I thought for symmetry it was worth adding [`intersection` and `union`], although I am not against removing them from this PR if it makes it more likely to be accepted

Yes please. Despite them being aliases for the already-defined Array operations, as you note, I think I'd prefer to keep those names free just in case we ever implement full-scale SQL set operations on Relation. `intersect?` seems fine, because its return value remains the same regardless of how we generate the answer -- but if we define `Relation#union(Relation) -> Array` now, it would be extremely difficult to later change it to returning a Relation.

john-h-k commented:
Have done that now @matthewd

skipkayhil reviewed (commented):
Can you [squash your commits](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#squashing-commits) as well?

john-h-k committed:
Add `intersects?` to Relation

Ruby 3.1 added `intersects?` which is equivalent to `(a & b).any?`. Rubocop added a corresponding cop, `Style/ArrayIntersect`, which transforms the old style to use `intersect?`. Unfortunately as `intersects?` is not delegated on `CollectionProxy`, this leads to false positives that need to be disabled for no good reason other than the fact the method isn't delegated.

This PR add delegation of `intersects?` to `Relation` which fixes this.

john-h-k commented:
Squashed

brandondrew commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,11 @@
+*   Add support for `Array#intersect?` to `ActiveRecord::Relation`.
```
This file uses `intersect?` (without an 'S'), but the other files use `intersects?` (with an 'S').  Perhaps that can be corrected, to help avoid bugs when people copy the method name from the change log?

byroot replied:
Good catch. Fixed in https://github.com/rails/rails/commit/e446ef82c8174a4bd6c98f670912bb395b3091d6, thank you.

guilleiguaran replied:
The CHANGELOG is correct, I fixed the code in https://github.com/rails/rails/commit/119b4b2af1d848362a1a9c1cd8a2bea46700b154
